File: 13 Dynamic/Java/Fibonacci.java
import java.util.HashMap;
import java.util.Map;

public class Fibonacci {

	Map<Integer, Integer> memory = new HashMap<>();

	int getnthinFiboSeriesDP(int n) {
		if (memory.containsKey(n)) {
			return memory.get(n);
		}

		else {
			if (n < 2) {
				memory.put(n, 1);
				return 1;
			}
			int result = getnthinFiboSeriesDP(n - 1) + getnthinFiboSeriesDP(n - 2);
			memory.put(n, result);
			return result;
		}
	}

	int printFibo(int n) {
		if (n == 1 || n == 0) {
			return 1;
		}
		return printFibo(n - 1) + printFibo(n - 2);
	}

	public static void main(String[] args) {
		Fibonacci obj = new Fibonacci();
		System.out.println(obj.getnthinFiboSeriesDP(5));
	}
}

File: 13 Dynamic/C#/Fibonacci.cs
using System.Collections.Generic;

namespace DP
{
    public class Fibonacci
    {
        /// <summary>
        /// Get Fibonacci number of n th element using recursion(no DP)
        /// 0th element is 1
        /// 1st element is 1
        /// 2nd element is 2
        /// so on..
        /// </summary>
        /// <param name="n">n th element</param>
        /// <returns></returns>
        public long GetFibonacci(int n)
        {
            if (n < 2)
            {
                return 1;
            }

            return GetFibonacci(n - 1) + GetFibonacci(n - 2);
        }

        Dictionary<int, long> memory = new Dictionary<int, long>();
        
        /// <summary>
        /// Fibonacci number of n th element recursively with memorization (Dynamic programming) in key value pair
        /// Memorization with Top down approach
        /// </summary>
        /// <param name="n">n th element</param>
        /// <returns></returns>
        public long GetFibonacciwithDP(int n)
        {
            long result;
            if (memory.ContainsKey(n))
            {
                long value;
                memory.TryGetValue(n, out value);
                return value;
            }

            else    // not required
            {
                if (n < 2)
                {
                    memory.Add(n, 1);
                    return 1;
                }

                result = GetFibonacciwithDP(n - 1) + GetFibonacciwithDP(n - 2);
                memory.Add(n, result);
            }
            return result;
        }

        /// <summary>
        /// Fibonacci number of n th element recursively with memorization (Dynamic programming) in array
        /// Memorization with Top down approach
        /// </summary>
        /// <param name="n">n th element</param>
        /// <returns></returns>
        public long GetFibonacciWithDPWithoutDictionary(int n)
        {
            long[] memory3 = new long[n];
            for (int i = 0; i < n; i++)
            {
                memory3[i] = -1;
            }

            return FibonacciPrivate(n - 1, memory3);
        }

        #region private methods

        private long FibonacciPrivate(int n, long[] memory3)
        {
            if (memory3[n] != -1)
            {
                return memory3[n];
            }
            else
            {
                if (n < 2)
                {
                    memory3[n] = 1;
                    return 1;
                }
                else
                {
                    long result = FibonacciPrivate(n - 1, memory3) + FibonacciPrivate(n - 2, memory3);
                    memory3[n] = result;
                    return result;
                }
            }

        } 
        #endregion
    }
}